packages/react/src/components/hooks/usePermissions.ts

```diff
@@ -0,0 +1,151 @@
+  "room.self.deaf",
+  "room.self.undeaf",
+  "room.self.screenshare",
+  "room.self.set_input_sensitivity",
+  "room.self.set_input_volume",
+  "room.self.set_output_volume",
+  "room.self.video_mute",
+  "room.self.video_unmute",
+  "room.self.set_position",
+  "room.member.set_position",
+  "room.self.set_meta",
+  "room.member.set_meta",
+];
+const defaultPermissions = [
+  "room.self.audio_mute",
+  "room.self.audio_unmute",
+  "room.self.video_mute",
+  "room.self.video_unmute",
+  "room.self.deaf",
+  "room.self.undeaf",
+  "room.self.set_input_volume",
+  "room.self.set_output_volume",
+  "room.self.set_input_sensitivity",
+  "room.self.screenshare",
+  "room.self.additional_source",
+];
+
+function permissionObjectFromPermissions(permissions: string[]) {
+  const [self, member] = ["room.self.", "room.member."].map(
+    (prefix: any): any => {
+      const [allPrefixPermissions, allowedPrefixPermissions] = [
+        allPermissions,
+        permissions,
+      ].map((list) =>
+        list
+          .filter((p) => p.startsWith(prefix))
+          .map((p) => p.substring(prefix.length))
+      );
+      const hasPermissions: any = {};
+      allPrefixPermissions.forEach(
+        (perm) =>
+          (hasPermissions[perm] = allowedPrefixPermissions.includes(perm))
+      );
+      hasPermissions.full = Object.keys(hasPermissions).every(
+        (v: string) => hasPermissions[v]
+      );
+      return hasPermissions;
+    }
+  );
+
+  const layout = {
+    list_available_layouts: permissions.includes("room.list_available_layouts"),
+    set_layout: permissions.includes("room.set_layout"),
+    full: false,
+  };
+  layout.full = layout.list_available_layouts && layout.set_layout;
+
+  let permissionObj: any = {
+    default:
+      defaultPermissions.every((p) => permissions.includes(p)) &&
+      Object.keys(self).length === defaultPermissions.length,
+    self,
+    member,
+    layout,
+    playback: permissions.includes("room.playback"),
+    recording: permissions.includes("room.recording"),
+    show_video_muted: permissions.includes("room.show_video_muted"),
+    hide_video_muted: permissions.includes("room.hide_video_muted"),
+    set_position: permissions.includes("room.set_position"),
+    set_meta: permissions.includes("room.set_meta"),
+  };
+
+  ["self", "member"].forEach((type) => {
+    permissionObj[type].audio = {
+      mute: permissionObj[type].audio_mute,
+      unmute: permissionObj[type].audio_unmute,
+      full: permissionObj[type].audio_mute && permissionObj[type].audio_unmute,
+    };
+    permissionObj[type].video = {
+      mute: permissionObj[type].video_mute,
+      unmute: permissionObj[type].video_unmute,
+      full: permissionObj[type].video_mute && permissionObj[type].video_unmute,
+    };
+    permissionObj[type].speaker = {
+      mute: permissionObj[type].deaf,
+      unmute: permissionObj[type].undeaf,
+      full: permissionObj[type].deaf && permissionObj[type].undeaf,
+    };
+  });
```

Comment:
I totally understand what you're saying. But I also see great value in the helper attributes being there, because more often than not, the meaningful unit of permission is 'audio_control' instead of 'audio_mute, but not audio_unmute'. 

First experiments with proxy almost reach there but not quite (https://codesandbox.io/s/eager-gauss-pno3m6?file=/src/index.js). 

I'll see if I can combine Proxies with other code to make maintainable but expressive permission object, but for now, lets go with this implementation.

Reply:
is the Proxy code in this PR stable at the moment?

Reply:
Oh yes. it works as intended now. First I tried a recursive proxy-only approach, where `P.self.audio` would evaluate to a boolean AND of all it's members (`P.self.audio` = `P.self.audio.muted` AND `P.self.audio.unmuted`) but that I found was inherently impossible in the language (unless there's some other trick I'm missing).

Now I'm pre-building the object and only using proxy to return `false` where JS would otherwise return `undefined`.